aiida_optimize/engines/_convergence.py

```diff
@@ -0,0 +1,236 @@
+# -*- coding: utf-8 -*-
+
+# Author: Austin Zadoks <austin.zadoks@epfl.ch>
+"""
+Defines a basic and generic parameter convergence engine.
+"""
+
+import itertools
+import typing as ty
+
+import numpy as np
+from aiida import orm
+from aiida_optimize.engines._base import (OptimizationEngineImpl, OptimizationEngineWrapper)
+from aiida_optimize.engines._result_mapping import Result
+
+__all__ = ['Convergence']
+
+
+class _ConvergenceImpl(OptimizationEngineImpl):
+    """
+    Implementation class for the convergence engine.
+    """
+    def __init__(
+        self,
+        *,
+        input_values: ty.Iterable[ty.Any],
+        tol: float,
+        input_key: str,
+        result_key: str,
+        convergence_window: int,
+        array_name: ty.Optional[str],
+        current_index: int,
+        result_values: ty.List[ty.Any],
+        initialized: bool,
+        logger: ty.Optional[ty.Any],
+        result_state: ty.Optional[ty.Dict[int, Result]] = None,
+    ):
+        super(_ConvergenceImpl, self).__init__(logger=logger, result_state=result_state)
+        self.input_values = input_values
+        self.tol = tol
+        self.input_key = input_key
+        self.result_key = result_key
+        self.array_name = array_name
+        self.convergence_window = convergence_window
+        self.current_index = current_index
+        self.result_values = result_values
+        self.initialized = initialized
+
+    @property
+    def _state(self) -> ty.Dict[str, ty.Any]:
+        """
+        Generate the engine state, including variables
+            input_values, tol, input_key, result_key, convergence_window,
+            current_index, result_values, initialized
+        and excluding variables
+            logger
+        """
+        return {k: v for k, v in self.__dict__.items() if k not in ['_result_mapping', '_logger']}
+
+    @property
+    def _result_window(self) -> ty.List[ty.Any]:
+        """
+        Create a list of results corresponding to the current convergence window
+        and convert those results to python / numpy objects from AiiDA objects
+        """
+        result_window = self.result_values[-self.convergence_window:]
+        for i, result in enumerate(result_window):
+            if isinstance(result, orm.ArrayData):
+                result_window[i] = result.get_array(self.array_name)
+            elif isinstance(result, (orm.Float, orm.Int)):
+                result_window[i] = result.value
+
+        return result_window
+
+    @property
+    def _distance_triangle(self) -> ty.List[ty.Any]:
```

Comment:
I guess the return type here could be `ty.List[ty.List[float]]`.

Reply:
Done!

Reply:
I'm not seeing any changes - did you push them?